[Xtensa] Lower GlobalAddress/BlockAddress/JumpTable

andreisfr opened the pull request:
This patch implements lowering of the GlobalAddress, BlockAddress, JumpTable and BR_JT. Also patch adds legal support of the BR_CC operation for i32 type.

andreisfr committed:
[Xtensa] Lower GlobalAddress/BlockAddress/JumpTable

This patch implements lowering of the GlobalAddress, BlockAddress, JumpTable
and BR_JT. Also patch adds legal support of the BR_CC operation for i32 type.

andreisfr commented:
@arsenm , @MaskRay , @s-barannikov , PTAL to this PR whenever you have the time.

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -42,6 +43,17 @@ getModifierVariantKind(XtensaCP::XtensaCPModifier Modifier) {
+        *OutStreamer,
+        MCInstBuilder(Xtensa::JX).addReg(MI->getOperand(0).getReg()));
+    return;
+  }
```
Needs a default?

andreisfr replied:
I created "default" case

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -52,16 +64,25 @@ void XtensaAsmPrinter::emitMachineConstantPoolValue(
 
-  if (XtensaSym->isPrivateLinkage())
-    SymName = ".L" + SymName;
+    MCSym = GetExternalSymbolSymbol(StringRef(SymName));
```
GetExternalSymbolSymbol uses Twine so you shouldn't need to build a new std::string here for the prefix. Also avoid hardcoding the prefix. Also, can't you just use getOrCreateSymbol? I think that takes care of the prefix?

andreisfr replied:
I substituted GetExternalSymbolSymbol to getOrCreateSymbol, hope I understood proposed solution correctly.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -519,6 +537,87 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+                                                 SelectionDAG &DAG) const {
+  const GlobalAddressSDNode *G = cast<GlobalAddressSDNode>(Op);
+  SDLoc DL(Op);
+  auto PtrVt = getPointerTy(DAG.getDataLayout());
```
Use the type of the incoming value instead of guessing it's an address space 0 pointer. You should hardly ever need getPointerTy

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -519,6 +537,87 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+  auto PtrVt = getPointerTy(DAG.getDataLayout());
+  const GlobalValue *GV = G->getGlobal();
+
+  // Check Op SDNode users
```
This handling doesn't make sense to me and you shouldn't depend on the user context. If you want to fold the global address into an immediate call operand, you can do that in your call handling

andreisfr replied:
Thank you very much for comment! I removed redundant code.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -519,6 +537,87 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+SDValue XtensaTargetLowering::LowerBlockAddress(BlockAddressSDNode *Node,
+                                                SelectionDAG &DAG) const {
+  const BlockAddress *BA = Node->getBlockAddress();
+  EVT PtrVT = getPointerTy(DAG.getDataLayout());
```
Use the type of the original node

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -519,6 +537,87 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+  const MachineJumpTableInfo *MJTI = MF.getJumpTableInfo();
+  SDValue TargetJT = DAG.getTargetJumpTable(JT->getIndex(), MVT::i32);
+  const DataLayout &TD = DAG.getDataLayout();
+  EVT PTy = getPointerTy(TD);
```
Use the type of the original node

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -519,6 +537,87 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+SDValue XtensaTargetLowering::LowerJumpTable(JumpTableSDNode *JT,
+                                             SelectionDAG &DAG) const {
+  SDLoc DL(JT);
+  EVT PtrVt = getPointerTy(DAG.getDataLayout());
```
Use the type of the original node

andreisfr replied:
fixed

andreisfr committed:
[Xtensa] Minor fixes in lowering GlobalAddress/BlockAddress/JumptTable.

Remove redundant code and correct pointer type variable initialization.
Also improve asm printing.

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -42,26 +43,48 @@ getModifierVariantKind(XtensaCP::XtensaCPModifier Modifier) {
+
+    if (XtensaSym->isPrivateLinkage()) {
+      const DataLayout &DL = getDataLayout();
+      MCSym = OutContext.getOrCreateSymbol(Twine(DL.getPrivateGlobalPrefix()) +
+                                           SymName);
```
I think something further down is supposed to append the prefix for you but I'm no MC expert

andreisfr replied:
I've found that that "DL.getPrivateGlobalPrefix()" code returns appropriate prefix for Xtensa for private linkage case and such code is used for similar purposes in ARM backend and other backends. So, I used it to substitute hardcoded ".L" prefix.
